File: frontend/src/utils/quantities.ts
const toNumber = (value: number | string | null | undefined): number => {
  const numeric = Number(value);
  return Number.isFinite(numeric) ? numeric : 0;
};

export const roundUpQuantity = (value: number | string | null | undefined): number => {
  const numeric = toNumber(value);
  if (numeric <= 0) return 0;
  return Math.ceil(numeric - 1e-9);
};

export const roundUpSignedQuantity = (value: number | string | null | undefined): number => {
  const numeric = toNumber(value);
  if (Math.abs(numeric) < 1e-9) return 0;
  if (numeric > 0) return Math.ceil(numeric - 1e-9);
  return -Math.ceil(Math.abs(numeric) - 1e-9);
};

export const formatQuantity = (value: number | string | null | undefined): string =>
  String(roundUpQuantity(value));

export const formatSignedQuantity = (value: number | string | null | undefined): string => {
  const numeric = roundUpSignedQuantity(value);
  return `${numeric >= 0 ? '+' : ''}${numeric}`;
};

File: backend/app/core/quantities.py
from __future__ import annotations

import math


def round_up_quantity(value: float | int | None) -> int:
    if value is None:
        return 0
    numeric = float(value)
    if numeric <= 0:
        return 0
    return int(math.ceil(numeric - 1e-9))


def round_up_signed_quantity(value: float | int | None) -> int:
    if value is None:
        return 0
    numeric = float(value)
    if abs(numeric) < 1e-9:
        return 0
    if numeric > 0:
        return int(math.ceil(numeric - 1e-9))
    return -int(math.ceil(abs(numeric) - 1e-9))
